core/src/eval/operation.rs

```diff
@@ -2418,28 +2406,43 @@ impl<R: ImportResolver, C: Cache> VirtualMachine<R, C> {
                             })
                             .collect();
 
+                        let ctrs_right_empty = ctrs_right_sieve.iter().all(Option::is_none);
                         let ctrs_right_dedup = ctrs_right_sieve.into_iter().flatten();
 
-                        ts.extend(ts1.into_iter().map(|t| {
-                            RuntimeContract::apply_all(t, ctrs_left_dedup.iter().cloned(), pos1)
-                                .closurize(&mut self.cache, env1.clone())
-                        }));
+                        let ctrs_left_empty = ctrs_left_dedup.is_empty();
+
+                        let arr = if ctrs_right_empty && ctrs_left_empty {
+                            ts1.extend(ts2);
+                            ts1
+                        } else if ctrs_left_empty {
+                            ts1.extend(ts2.into_iter().map(|t| {
+                                RuntimeContract::apply_all(t, ctrs_right_dedup.clone(), pos1)
+                                    .closurize(&mut self.cache, env1.clone())
+                            }));
+                            ts1
+                        } else {
+                            let mut ts = Array::default();
+
+                            ts.extend(ts1.into_iter().map(|t| {
+                                RuntimeContract::apply_all(t, ctrs_left_dedup.iter().cloned(), pos1)
+                                    .closurize(&mut self.cache, env1.clone())
+                            }));
 
```

Comment:
I didn't understand this, sorry. Are you referring to the pattern of making an `Array` and then extending it twice? It isn't written that way as an optimization, but because the (more natural, IMO) approach of doing a `chain` and then `collect` fails the borrow checker: both closures want a `&mut self.cache`

Reply:
My bad, I thought this was some kind of optimization of array concatenation when the left side or the right side is empty, but it's more than that - it's about contract deduplication and there is additional `apply_all` in the `else` case. You can just ignore my previous comment, sorry.